### PR TITLE
Support DSN in Client constructor for config argument (#1640)

### PR DESCRIPTION
**Pull request by alamirault**
This PR allow to pass DSN to the Elastica\Client constructor as discussed in #1640 .

For example: `$client = new Client('https://user:p4ss@foo.com:9200?persistent=false&retryOnConflict=2');`

I have just a doubt about `url` config parameter. Should I build it from `transport`, `host`, `port` and `path` ? Or it works when these args are filled ?

Another question: Is there a doc to update ? Elastica.io appears to be outdated.

### CODE REVIEW

File: lib/Elastica/Dsn.php [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        $this->connections = $connections;
+        $this->roundRobin = $roundRobin;
+        $this->log = $log;
+        $this->retryOnConflict = $retryOnConflict;
+        $this->bigintConversion = $bigintConversion;
+        $this->username = $username;
+        $this->password = $password;
+    }
+
+    public static function fromString(string $dsn): self
```
Comment (ruflin) [inline lib/Elastica/Dsn.php]: As all params are kind of hardcoded, how do we keep it up-to-date on changes?

Comment (alamirault) [reply lib/Elastica/Dsn.php]: What changes ? If you talk about config array maybe we have to move Dsn object to a Configuration object which have methods `fromDsn` and `fromArray` and use it in the Client.

**Comment by alamirault** [comment]
I reworked my PR. Now the client configuration is exported in new class. It avoid code duplication.
I also allow any extra option (removed hardcoded part)

**Comment by ruflin** [comment]
I like the change to use a separate class for the config object. I'm trying to poke holes into this change to see what side effects it could have. So far the only thing I could think of is that something that inherits the client could mess around with `_config` but even this should be fine as it can still be an array.

It seems everything should continue to works as before?

**Comment by alamirault** [comment]
I think is only side effect. Indeed everything works without any changes.

I also would like deprecate `setConfig`,`getConfig`,`setConfigValue`,`getConfigValue` and add a ClientConfiguration getter in Client. WDYT ?

**Comment by ruflin** [comment]
@alamirault Your proposal SGTM. I already merged the PR as I think these additional changes / deprecations should go into a separate PR. I hope you are ok with that?

**Comment by alamirault** [comment]
I agree. I will make another PR when I have time.

**Pull request merged**